Default dialect when no schema specified

diyaayay opened the pull request:
Resolves #6 #4 

Changes:

- Setup the server to manage configuration. 
- Server now responds to configuration change events from the client.
- Assigns a default dialect when no `$schema` keyword is used.

Video demonstrating the changes:
**VS Code** :


https://github.com/hyperjump-io/json-schema-language-tools/assets/110971977/db53dadc-9388-440c-87dd-4c0ef1191804

**Neovim** :

https://github.com/hyperjump-io/json-schema-language-tools/assets/110971977/14c45890-00ff-484a-9bae-fbcfe3a5f581

Thanks to my partner, @wthrajat ,  for his invaluable assistance throughout our collaboration. His unwavering support, dedication, and assistance in code debugging, guidance in Neovim testing, have significantly enhanced our workflow efficiency and productivity.

**This PR has been created as a part of GSoC qualification task.**

jdesrosiers commented on language-server/src/server.js:
```diff
@@ -72,6 +78,9 @@ connection.onInitialize(({ capabilities, workspaceFolders }) => {
 
 connection.onInitialized(async () => {
+  if (hasConfigurationCapability) {
+    connection.client.register(DidChangeConfigurationNotification.type, undefined);
```
It's odd to explicitly pass `undefined` to a function. Unless you have a good reason, let's remove that parameter.

diyaayay replied:
I removed it and tested, works without this parameter too. My bad, I took a reference from how it's done in vscode lsp extension.

jdesrosiers commented on language-server/src/server.js:
```diff
@@ -72,6 +78,9 @@ connection.onInitialize(({ capabilities, workspaceFolders }) => {
 });
 
 connection.onInitialized(async () => {
+  if (hasConfigurationCapability) {
+    connection.client.register(DidChangeConfigurationNotification.type, undefined);
+  }
```
Code Style: I generally like to have whitespace after a block.

```suggestion
  if (hasConfigurationCapability) {
    connection.client.register(DidChangeConfigurationNotification.type, undefined);
  }
  
```

diyaayay replied:
Will keep it in mind from next time and double check it for now for all the places this might have occurred.

jdesrosiers commented on language-server/src/server.js:
```diff
@@ -107,6 +116,30 @@ connection.onInitialized(async () => {
   await validateWorkspace();
 });
 
+const documentSettings = new Map(); // Consider a Map for cache
```
Useless comment.
```suggestion
const documentSettings = new Map();
```

jdesrosiers commented on language-server/src/server.js:
```diff
@@ -107,6 +116,30 @@ connection.onInitialized(async () => {
 
+const documentSettings = new Map(); // Consider a Map for cache
+let globalSettings = {
+  enableDetailedErrors: false,
+  schemaValidationSeverity: "error",
```
These don't belong. You can remove them. 

```suggestion
```

jdesrosiers commented on language-server/src/server.js:
```diff
@@ -107,6 +116,30 @@ connection.onInitialized(async () => {
+let globalSettings = {
+  enableDetailedErrors: false,
+  schemaValidationSeverity: "error",
+  defaultDialect: "http://json-schema.org/draft-07/schema#"
```
There shouldn't be a default value for this config. A default dialect should only apply if the user explicitly sets a default in their config.

```suggestion
```

jdesrosiers commented on language-server/src/server.js:
```diff
@@ -107,6 +116,30 @@ connection.onInitialized(async () => {
+  schemaValidationSeverity: "error",
+  defaultDialect: "http://json-schema.org/draft-07/schema#"
+}; // Sensible defaults
+
```
Code style: Too much whitespace.

```suggestion
```

jdesrosiers commented on language-server/src/server.js:
```diff
@@ -148,43 +179,70 @@ documents.onDidChangeContent(async ({ document }) => {
   const diagnostics = [];
-
+  const settings = await getDocumentSettings(document.uri);
+  connection.console.log(JSON.stringify(settings));
```
While I'm sure this was useful in debugging, it should be removed before this is merged.

```suggestion
```

jdesrosiers commented on language-server/src/server.js:
```diff
@@ -148,43 +179,70 @@ documents.onDidChangeContent(async ({ document }) => {
   }
 
+
+  connection.console.log(`Schema used: ${dialectUri}`);
```
While I'm sure this was useful in debugging, it should be removed before this is merged.

```suggestion
```

jdesrosiers commented on language-server/src/server.js:
```diff
@@ -148,43 +179,70 @@ documents.onDidChangeContent(async ({ document }) => {
+  }
+
+  // Revalidate all open text documents
+  documents.all().forEach((document) => validateSchema(document));
```
We're going to need to revalidate the entire workspace, not just the open documents. Look for the `validateWorkspace` function and use that instead.

diyaayay replied:
Ahh yes, I overlooked this. makes much more sense to revalidate the entire workspace.

jdesrosiers commented on vscode/package.json:
```diff
@@ -31,6 +31,17 @@
+        "properties": {
+          "jsonSchemaLanguageServer.defaultDialect": {
+            "type": "string",
+            "description": "The default JSON Schema dialect to use if none is specified in the schema document",
+            "default": "http://json-schema.org/draft-07/schema#"
```
There should be no default.

```suggestion
            "description": "The default JSON Schema dialect to use if none is specified in the schema document"
```

jdesrosiers commented on vscode/package.json:
```diff
@@ -31,6 +31,17 @@
+  "contributes": {                       
+    "configuration": {
+      "title": "Configuration for JSON Schema Language Server",
+        "properties": {
+          "jsonSchemaLanguageServer.defaultDialect": {
+            "type": "string",
+            "description": "The default JSON Schema dialect to use if none is specified in the schema document",
+            "default": "http://json-schema.org/draft-07/schema#"
+          }
+        }
+      }
+    }
   }
```
There's a problem with your indentation.

```suggestion
      "properties": {
        "jsonSchemaLanguageServer.defaultDialect": {
          "type": "string",
          "description": "The default JSON Schema dialect to use if none is specified in the schema document",
          "default": "http://json-schema.org/draft-07/schema#"
        }
      }
    }
  }
}
```

diyaayay replied:
I guess this happened because of disabling vscode's inbuilt LSP, I'll fix this.

jdesrosiers reviewed (changes requested):
I did a first pass review, but you need to rebase to the latest commits on "main" before I can do a full review.

The new config must not have a default value. The default behavior should be that users need to use `$schema` to declare their intended dialect. They need to explicitly set the config to get the "defaultDialect" behavior.

In both demos, you never set the config. You only show what happens when the config's default gets used. Show the behavior without the config set, then set it and show the behavior with the config set.

In the neovim demo, at the end, the language server crashes. You'll need to fix that.

jdesrosiers reviewed (changes requested):
It's looking good so far. I've left a few (mostly cosmetic) suggestions inline, but I have a couple more things I noticed.

I haven't figured out how best to organize the code yet. There's not a lot of organization, but there is some. I'm trying to keep related functionality grouped as well as possible. You'll see each feature headed by a comment in all caps marking the section where that feature is implemented. The exception to this grouping is initialization stuff because that all has to be in the same place. I'd like to see a section for `// CONFIGURATION` with all the configuration-related code (other than initialization) under that heading. That would help keep things somewhat organized until I can come up with a better organization strategy.

I did find one bug. The semantic tokens functionality isn't working when the defaultDialect is used. Oddly, I'm not seeing semantic tokens working at all in the vscode demo, but you can see what I'm talking about in the neovim demo. The keywords are highlighted when `$schema` is used, but that disappears when `$schema` is removed and the defaultDialect applies. This should be simple to fix, just use the config in the semantic tokens functionality the same way you use it for the schema validation functionality.

diyaayay commented on vscode/package.json:
```diff
@@ -31,6 +31,16 @@
     "onLanguage:json"
   ],
-  "contributes": {
+  "contributes": {                       
```
Didn't notice these. Prettier doesn't show whitespaces in package.json, fixing these manually now.

wthrajat replied:
Resolved by @diyaayay's commit 8518ad8

diyaayay commented on language-server/src/server.js:
```diff
@@ -169,7 +204,7 @@ const validateSchema = async (document) => {
     if (!output.valid) {
       for await (const [instance, message] of invalidNodes(output)) {
-        diagnostics.push(buildDiagnostic(instance, message));
+        diagnostics.push(buildDiagnostic(instance, message, settings.schemaValidationSeverity === "warning" ? DiagnosticSeverity.Warning : DiagnosticSeverity.Error));
```
Yes, should be reverted to the parameters being used previously.

diyaayay commented:
> It's looking good so far. I've left a few (mostly cosmetic) suggestions inline, but I have a couple more things I noticed.
> 
> I haven't figured out how best to organize the code yet. There's not a lot of organization, but there is some. I'm trying to keep related functionality grouped as well as possible. You'll see each feature headed by a comment in all caps marking the section where that feature is implemented. The exception to this grouping is initialization stuff because that all has to be in the same place. I'd like to see a section for `// CONFIGURATION` with all the configuration-related code (other than initialization) under that heading. That would help keep things somewhat organized until I can come up with a better organization strategy.
> 
> I did find one bug. The semantic tokens functionality isn't working when the defaultDialect is used. Oddly, I'm not seeing semantic tokens working at all in the vscode demo, but you can see what I'm talking about in the neovim demo. The keywords are highlighted when `$schema` is used, but that disappears when `$schema` is removed and the defaultDialect applies. This should be simple to fix, just use the config in the semantic tokens functionality the same way you use it for the schema validation functionality.

The sematic tokens functionality seems to be theme dependent, as mentioned here too https://code.visualstudio.com/api/language-extensions/semantic-highlight-guide#:~:text=Semantic%20tokenization%20allows%20language%20servers,the%20syntax%20highlighting%20from%20grammars.
I changed the theme and it seems to work fine. It will be visible in the updated vscode demo. So, I guess we don't need to fix it?

diyaayay commented on language-server/src/server.js:
```diff
@@ -149,9 +178,15 @@ documents.onDidChangeContent(async ({ document }) => {
   const diagnostics = [];
 
+  let settings = await getDocumentSettings(document.uri);
+  if (!settings) {
+    connection.console.log("Settings not available");
+    settings = {};
+  }
```
Setting it to `globalSettings` would be the correct way of assigning to `settings`. We'll handle it in `getDocumentSettings` to avoid duplication.

jdesrosiers replied:
@wthrajat Sure, right now there's no difference, but my understanding is that `globalSettings` is initialized to the default configs. Currently that's empty because we only have one config and it doesn't have a default, but that's not always going to be true. At some point I expect we will have more configs and some of them might have defaults.

So, the question is, when `connection.workspace.getConfiguration` returns nothing, is the desired behavior that the settings should be empty, or should it be the default configs from `globalSettings`?

As a corollary to that point, should the `globalSettings` get merged into the result from `connection.workspace.getConfiguration` to fill in default for any potentially missing configs that need defaults?

diyaayay replied:
As far as my understanding of `globalSettings`, it should be that The global settings are used when the `workspace/configuration` request is not supported by the client. So, I think when `connection.workspace.getConfiguration`
returns nothing, let `settings` be an empty object but when configuration only is not supported by the client then we use the `globalSettings`. Does that seem fine? @wthrajat @jdesrosiers

diyaayay replied:
So, I think `globalSettings` being merged into the result depends on the behavior we expect of `globalSettings` as I mentioned above. But what if the user doesn't want globalSettings to be merged, like what if they just want the settings that they have specified in `settings.json` . I maybe wrong but if this might be the case then I don't think `result` and the potentially missing configs should be merged.

jdesrosiers replied:
Thanks for the thoughtful analysis @diyaayay. It sounds like merging isn't the right thing. However, I disagree with the conclusion that `getConfiguration` returning nothing is the same as an empty object. I think an empty object is the user explicitly saying they want no settings set (which is why merging is probably the wrong choice), but returning nothing is the user saying nothing about what settings they want. If the user says nothing about what settings they want, I think it makes sense that they get the `globalSettings`.

diyaayay replied:
resolved with https://github.com/hyperjump-io/json-schema-language-tools/pull/25/commits/53c47a4e381f28f2d8f139c14e890e8711ae0844

wthrajat commented on language-server/src/server.js:
```diff
@@ -133,7 +133,7 @@ async function getDocumentSettings(resource) {
   }
 
-  return result;
+  return result ?? {};
```
@jdesrosiers Now we're returning an empty object when settings is `null`. This is done to ensure that in `neovim` the server doesn't crashes when settings is `null`.
Tested in both VSCode and Neovim.

jdesrosiers replied:
Although this works, wouldn't it be cleaner to never put a `null` in the `documentSettings` Map in the first place?

diyaayay replied:
yes, so maybe storing `result` in `documentSettings` only when it is not null should be the solution.

wthrajat replied:
Yes that'd be better. Solved by @diyaayay in this commit 93372c1

jdesrosiers commented:
@diyaayay 

> I changed the theme and it seems to work fine. Does it?

From your screenshot, it looks like you have a theme that supports semantic tokens now. Since the declared dialect (`""`) is unrecognized, the only keyword that's highlighted is `$schema`. If you put a supported dialect in `$schema`, you should see more keywords highlighted based on the dialect you used. If you remove `$schema` and have defaultDialect set to a supported dialect, then you should see keywords highlighted based on the default dialect. That last part is what isn't working correctly yet.

wthrajat commented on vscode/package.json:
```diff
@@ -31,6 +31,16 @@
+    }
   }
 }
+
```
Resolved by 8518ad8

diyaayay commented:
> @diyaayay
> 
> > I changed the theme and it seems to work fine. Does it?
> 
> From your screenshot, it looks like you have a theme that supports semantic tokens now. Since the declared dialect (`""`) is unrecognized, the only keyword that's highlighted is `$schema`. If you put a supported dialect in `$schema`, you should see more keywords highlighted based on the dialect you used. If you remove `$schema` and have defaultDialect set to a supported dialect, then you should see keywords highlighted based on the default dialect. That last part is what isn't working correctly yet.

I've updated the recording of how the highlighting works now after updating the code. We tried other ways too by simply doing it the way ` contextDialectUri ` is assigned in `validateSchema` but it showed vague behaviour in highlighting, so the current approach seemed to be showing the results.

![WhatsApp Image 2024-03-16 at 01 23 24_2071232c](https://github.com/hyperjump-io/json-schema-language-tools/assets/110971977/f8a56523-c223-4229-8c6d-6e1e147663aa)


> It's looking good so far. I've left a few (mostly cosmetic) suggestions inline, but I have a couple more things I noticed.
> 
> I haven't figured out how best to organize the code yet. There's not a lot of organization, but there is some. I'm trying to keep related functionality grouped as well as possible. You'll see each feature headed by a comment in all caps marking the section where that feature is implemented. The exception to this grouping is initialization stuff because that all has to be in the same place. I'd like to see a section for `// CONFIGURATION` with all the configuration-related code (other than initialization) under that heading. That would help keep things somewhat organized until I can come up with a better organization strategy.
> 
> I did find one bug. The semantic tokens functionality isn't working when the defaultDialect is used. Oddly, I'm not seeing semantic tokens working at all in the vscode demo, but you can see what I'm talking about in the neovim demo. The keywords are highlighted when `$schema` is used, but that disappears when `$schema` is removed and the defaultDialect applies. This should be simple to fix, just use the config in the semantic tokens functionality the same way you use it for the schema validation functionality.

I've tried putting all the configuration code that we've added under one section also keeping in mind how the other sections are organised , let me know if this structure doesn't work or if it should be some other way

jdesrosiers reviewed (changes requested):
It looks like [this](https://github.com/hyperjump-io/json-schema-language-tools/pull/25#discussion_r1525298984) and [this](https://github.com/hyperjump-io/json-schema-language-tools/pull/25#discussion_r1525309722) haven't been addressed yet. Other than that, the code looks good.

I notice in the vscode that you're getting a lot of errors pop up. Is that related to what you were talking about [here](https://github.com/hyperjump-io/json-schema-language-tools/pull/25#issuecomment-2000364202)? The highlighting seemed to be working fine otherwise. I could be wrong, but I vaguely remember that the semantic tokens handler couldn't take an async function. That could be related to the problem you're facing.

In the neovim demo, you never set the config to show the behavior when the config is set. You're only showing me the existing functionality, not the new functionality.

jdesrosiers commented on language-server/src/server.js:
```diff
@@ -135,6 +144,40 @@ connection.listen();
+    return globalSettings;
+  }
+
+  let result = documentSettings.get(resource);
+  if (!result) {
+    result = await connection.workspace.getConfiguration({
+      scopeUri: resource,
+      section: "jsonSchemaLanguageServer"
+    });
+    if (result !== null) {
+      documentSettings.set(resource, result);
+    }
+  }
+
+  return result ?? {};
```
Actually, I was think more like applying the default before adding it to the Map. This feels more straightforward.

```suggestion
  if (!documentSettings.has(resource)) {
    const result = await connection.workspace.getConfiguration({
      scopeUri: resource,
      section: "jsonSchemaLanguageServer"
    });
    documentSettings.set(resource, result ?? globalSettings);
  }

  return documentSettings.get(resource);
```

diyaayay commented:
> It looks like [this](https://github.com/hyperjump-io/json-schema-language-tools/pull/25#discussion_r1525298984) and [this](https://github.com/hyperjump-io/json-schema-language-tools/pull/25#discussion_r1525309722) haven't been addressed yet. Other than that, the code looks good.
> 
> I notice in the vscode that you're getting a lot of errors pop up. Is that related to what you were talking about [here](https://github.com/hyperjump-io/json-schema-language-tools/pull/25#issuecomment-2000364202)? The highlighting seemed to be working fine otherwise. I could be wrong, but I vaguely remember that the semantic tokens handler couldn't take an async function. That could be related to the problem you're facing.
> 
> In the neovim demo, you never set the config to show the behavior when the config is set. You're only showing me the existing functionality, not the new functionality.

![image](https://github.com/hyperjump-io/json-schema-language-tools/assets/110971977/ca77fdb0-8b9e-4b13-93d7-9cde538412d7)

@jdesrosiers 
Should this be fine? The errors don't pop up anymore, I've pushed the code.

jdesrosiers commented:
The demos look good! It looks like your branch is out of date and has some conflicts. Please rebase and resolve conflicts before I do my next review.

jdesrosiers commented on language-server/src/server.js:
```diff
@@ -135,6 +144,37 @@ connection.listen();
+  if (hasConfigurationCapability) {
+    documentSettings.clear();
+  } else {
+    globalSettings = change.settings.jsonSchemaLanguageServer || globalSettings;
```
This should probably use `??`, right?

diyaayay replied:
Ah, yes. I think we had a confusion about when to use global settings back then.

jdesrosiers commented on language-server/src/server.js:
```diff
@@ -259,9 +300,16 @@ const getTokenBuilder = (uri) => {
   const instance = JsoncInstance.fromTextDocument(document);
-  const dialectUri = instance.get("#/$schema").value();
+  const $schema = instance.get("#/$schema");
+  let dialectUri;
+  try {
+    dialectUri = $schema?.value() ?? settings.defaultDialect;
+  } catch (error) {
+    // SKIP
+  }
```
I think the error you were encountering here was caused by a bug that was fixed in JsoncInstance and when you rebased, you got the fix and this hack to catch the error was no longer necessary. The following works for me.

```suggestion
  const dialectUri = $schema.value() ?? settings.defaultDialect;
```

diyaayay replied:
Oh okay, This was the initial approach. But had to use the hack before rebasing, works now.

diyaayay replied:
I did commit the code but later realise that this is the semantic token highlighting when I do not use the hack:
![image](https://github.com/hyperjump-io/json-schema-language-tools/assets/110971977/87375dff-c10b-4d34-868a-b36f26f80694)


and below is the semantic highlighting when I use the hack.
![Uploading image.png…]()

diyaayay replied:
@jdesrosiers So, Should I revert to using the hack or try finding an alternative to semantic token highlighting.

jdesrosiers replied:
You have a typo. The `value` method is not being called. (You missed the `()`)

jdesrosiers reviewed (changes requested):
Almost there. I left a couple minor suggestions inline, but I found one bug in my exploratory testing.

If you set `defaultDialect` to an unsupported dialect (For example: `https://example.com/my-dialect`) and the schema doesn't have `$schema`, the error you get the "no dialect" error instead of the "unknown dialect" error.

diyaayay commented:
> Almost there. I left a couple minor suggestions inline, but I found one bug in my exploratory testing.
> 
> If you set `defaultDialect` to an unsupported dialect (For example: `https://example.com/my-dialect`) and the schema doesn't have `$schema`, the error you get the "no dialect" error instead of the "unknown dialect" error.

![image](https://github.com/hyperjump-io/json-schema-language-tools/assets/110971977/59890bc3-9170-420d-8067-d6c6c1ac54c6)

![image](https://github.com/hyperjump-io/json-schema-language-tools/assets/110971977/a8f4c1f5-82f6-49d2-a1c5-f53e8bac4b7c)

jdesrosiers commented on language-server/src/server.js:
```diff
@@ -149,21 +189,26 @@ documents.onDidChangeContent(async ({ document }) => {
         diagnostics.push(buildDiagnostic($schema, "Unknown dialect"));
       } else {
-        diagnostics.push(buildDiagnostic(schemaInstance, "No dialect"));
+        if (contextDialectUri) {
+          diagnostics.push(buildDiagnostic(schemaInstance, "Unknown dialect"));
+        } else {
+          diagnostics.push(buildDiagnostic(schemaInstance, "No dialect"));
+        }
```
The nesting is unnecessary.

This,

```javascript
if () {
} else {
  if () {
  } else {
  }
}
```

can be this,

```javascript
if () {
} else if () {
} else {
}
```

jdesrosiers reviewed (approved):
:tada:

jdesrosiers commented:
I merged manually because I wanted to squash commits, but it seems to have confused Github. I have to close this PR, but it was actually successfully merged. Thanks for your hard work!
